lib/Sema/CSSimplify.cpp

```diff
@@ -1410,6 +1410,22 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
     case ConstraintKind::ArgumentConversion:
     case ConstraintKind::OperatorArgumentTupleConversion:
     case ConstraintKind::OperatorArgumentConversion:
+      if (type1->is<InOutType>() && typeVar2 &&
+          typeVar2->getImpl().canBeInOut()) {
+        // Upgrade type2 to an InOutType
+        auto inOutType1 = type1->getAs<InOutType>();
+        
+        unsigned typeVar3Options = typeVar2->getImpl().getOptions();
+        typeVar3Options &= ~TVO_CanBeInOut;
```

Comment:
I think we should add `TVO_MustBeMaterializable` to `typeVar3Options`, because one cannot have an `inout` of a non-materializable type. (I'm not sure I understand why we need the new type variable at all... see below).